### PR TITLE
506755: Feature request: Buildship offline mode

### PR DESCRIPTION
**Pull request by rlagoue**
Task-Url: https://bugs.eclipse.org/bugs/show_bug.cgi?id=506755

@oehme please check this out...

### CODE REVIEW

**Comment by donat** [comment]
Thanks for the contribution @rlagoue, I think it makes sense to add this feature to Buildship. I'll review your changes shortly.

In the meanwhile, please check why the Eclipse IP validation failed for this PR (see above). The following document contains the requirements for contributions: https://www.eclipse.org/projects/handbook/#resources-cla

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/configuration/internal/DefaultWorkspaceConfigurationManager.java [hunk 1]
```diff
@@ -29,12 +29,14 @@
 public class DefaultWorkspaceConfigurationManager implements WorkspaceConfigurationManager {
 
     private static final String GRADLE_USER_HOME = "gradle.user.home";
+    private static final String GRADLE_OfflineMode = "gradle.offlineMode";
```
Comment (donat) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/configuration/internal/DefaultWorkspaceConfigurationManager.java]: GRADLE_OFFLINE_MODE = "gradle.offline.mode"

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/configuration/WorkspaceConfiguration.java [hunk 2]
```diff
@@ -20,13 +20,19 @@
...
         this.gradleUserHome = gradleUserHome;
+        this.gradleIsOffline = gradleIsOffline;
     }
 
     @Nullable
     public File getGradleUserHome() {
         return this.gradleUserHome;
     }
+
+    public boolean getGradleIsOffline() {
```
Comment (donat) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/configuration/WorkspaceConfiguration.java]: Change the method name to `isOfflineMode()`

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/i18n/CoreMessages.java [hunk 3]
```diff
@@ -60,6 +60,7 @@
     public static String ErrorMessage_0_AlreadyExists;
 
     public static String Preference_Label_GradleUserHome;
+    public static String Preference_Label_GradleOfflineMode;
```
Comment (donat) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/i18n/CoreMessages.java]: `Preference_Label_OfflineMode`

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/preferences/GradleWorkbenchPreferencePage.java [hunk 4]
```diff
@@ -55,6 +55,7 @@
     private final Validator<File> gradleUserHomeValidator;
 
     private Text gradleUserHomeText;
+    private Button gradleIsOfflineCheckbox;
```
Comment (donat) [inline org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/preferences/GradleWorkbenchPreferencePage.java]: `offlineModeCheckbox`

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/preferences/GradleWorkbenchPreferencePage.java [hunk 5]
```diff
@@ -70,12 +71,18 @@ protected Control createContents(Composite parent) {
...
         Group gradleUserHomeGroup = createGroup(page, CoreMessages.Preference_Label_GradleUserHome + ":");
         createGradleUserHomeSelectionControl(gradleUserHomeGroup);
+        createGradleIsOfflineCheckbox(page);
 
         initFields();
 
         return page;
     }
 
+    private void createGradleIsOfflineCheckbox(Composite page) {
```
Comment (donat) [inline org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/preferences/GradleWorkbenchPreferencePage.java]: `createOfflineModeCheckbox`

**Comment by rlagoue** [comment]
Hi Donat,

the changes you requested have been committed. Concerning the IP validation, I don't really know, what I have to do, to make it work.

BR

**Comment by donat** [comment]
Unfortunately the IP validation is mandatory for all Eclipse projects. But fear not, it is rather easy: just create an account on https://dev.eclipse.org with the same email address you use on GitHub. Then, when you log in do two things: sign the CLA (you will be prompted as soon as you log in), and in the  account settings set your GitHub ID. Finally, you have one thing left: modify your commits to have a signed-off-by entries at the end of your commits. Let me know if you need help on any of these steps.

**Comment by rlagoue** [comment]
Hi Donat,

- CLA signing ? done (did it the day of my first commit)
- Github ID on the eclipse settings page? done (just dit it)
- Commits modification to have signed-off-by entries? don't know how to dot that.

I need help for the last point
BR

**Comment by donat** [comment]
Here's how you can do it from the command line
```
git rebase -i head~2 # modify the last two commits
# in the editor change the keyword `pick` to `edit` and save
git commit --amend -s # adds the signed-off-by entry to your first commit 
git rebase --continue
git commit --amend -s # adds the signed-off-by entry to your second commit  
git rebase --continue
git log # verify manually if the signed-off-by entries are added
git diff head~2 # verify that your changes are there
git push -f your_forks_remote_repository bug-506755 # force-push changes to replace the original commits with the signed-offed ones
```

**Comment by rlagoue** [comment]
Hi Donat,

Should I just add "signed-off-by". I think my problem is that, I don't understand what this "signed-off-by" is all about.

**Comment by donat** [comment]
This is a [legal requirement](http://stackoverflow.com/questions/1962094/what-is-the-sign-off-feature-in-git-for), which the Eclipse Foundation takes very seriously.

**Comment by donat** [comment]
@rlagoue This is what it precisely means:

> By making and signing (by using the git signed-off-by mechanism or any other signing mechanism which may be specified by the Eclipse Foundation from time to time) a contribution to an Eclipse project, I certify that:
- I have authored 100% of the contribution.
- I have the necessary rights to submit this contribution, including any necessary permissions from my employer.
- I am providing this contribution under the license(s) associated with the Eclipse Foundation project I am contributing to.
- I understand and agree that Eclipse projects and my contributions are public, and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with the license(s) involved.

**Comment by rlagoue** [comment]
@donat 

Thank you for the explanation. I followed you very precise how to, and as we can see, everything is now ok.
Your support was great.

Thanks

**Comment by donat** [comment]
Thank you for taking the time for implementing this feature and for the patience for this whole legal stuff. If you have any suggestions what would have helped you - let it be more elaborate documentation or something else - please share it with us and I'll try to do something about it.

**Pull request merged**